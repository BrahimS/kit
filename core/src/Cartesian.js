import React, { Fragment } from 'react'

import { cartesianProduct } from './util'

export default ({ component, ...props }) => {
  const combinations = cartesianProduct(props)
  const Component = component

  return (
    <div>
      {combinations.map((props, i) =>
        <Component key={i} {...props} />
      )}
    </div>
  )
}
