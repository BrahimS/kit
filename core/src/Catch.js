import React from 'react'

class Catch extends React.Component {
  constructor() {
    super()

    this.state = {}
  }

  componentDidCatch(err) {
    this.setState({ err })
  }

  componentWillReceiveProps(next) {
    if (next.children !== this.props.children) {
      this.setState({ err: null })
    }
  }

  render() {
    const { err } = this.state

    if (err) return <pre>{err.toString()}</pre>

    return this.props.children
  }
}

export default Catch
