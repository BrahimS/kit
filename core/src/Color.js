import React from 'react'
import contrast from 'get-contrast'
import { Debug } from '.'
import { flatten } from './util'
import { Box, Code, Flex, Text } from './ui'

export const Swatch = ({ name, value, ...props }) => (
  <React.Fragment>
    <Box p={3} {...props}>
      <Box bg={value} p={[3, 4, 5]} />
      <Text>{name}</Text>
      <Code f={0}>{value}</Code>
    </Box>
  </React.Fragment>
)

export const Group = ({ colors, prefix, ...props }) => (
  <Flex {...props}>
    {colors.map((c, i) => {
      let name = null
      let value = null
      if (typeof c === 'object') {
        name = Object.keys(c)[0]
        value = c[name]
      } else {
        name = prefix ? `${prefix}.${i}` : c
        value = c
      }

      const color = contrast.isAccessible(value, 'white') ? 'white' : 'black'

      return (
        <Box key={name} bg={value} color={color} p={3}>
          <Text >{name}</Text>
        </Box>
      )
    })}
  </Flex>
)

export const Contrast = ({ color, bg, children }) => (
  <Box p={[3, 4, 5]} bg={bg} color={color}>
    <Text f={[5, 6, 128]} fontWeight='bold'>
      Aa
    </Text>
    <Flex alignItems='baseline'>
      <Code f={6}>{contrast.ratio(bg, color).toFixed(2)}</Code>
      <Text f={4} fontWeight='bold'>{contrast.score(bg, color)}</Text>
    </Flex>
    {children}
  </Box>
)

const Color = ({ colors }) => {
  const colorsObj = flatten(colors)
  const colorsArr = Object.keys(colorsObj).reduce((acc, key) =>
    acc.concat([{ key, name: key, value: colorsObj[key] }])
  , [])

  return (
    <Flex
      flexWrap='wrap'
      children={colorsArr.map(color =>
        <Swatch
          width={[1/2, 1/3, 1/5]}
          {...color}
        />
      )}
    />
  )
}

Color.Group = Group
Color.Swatch = Swatch
Color.Contrast = Contrast

export default Color
