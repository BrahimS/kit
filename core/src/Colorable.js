import React from 'react'
import colorable from 'colorable'

export default ({ colors, children, ...opts }) => children(colorable(colors, opts))
