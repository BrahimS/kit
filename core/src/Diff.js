import React, { Fragment } from 'react'
import nano from 'nano-style'
import { Box } from './ui'

const DiffBox = nano(Box)(({ blendMode }) => ({
  position: 'relative',
  mixBlendMode: blendMode || 'multiply'
}))

const ComponentBox = nano(Box)(({ blendMode }) => ({
  position: 'absolute',
  mixBlendMode: blendMode || 'multiply'
}))

export default ({ blendMode, a, b }) => (
  <DiffBox>
    <ComponentBox blendMode={blendMode}>{a}</ComponentBox>
    <ComponentBox blendMode={blendMode}>{b}</ComponentBox>
  </DiffBox>
)
