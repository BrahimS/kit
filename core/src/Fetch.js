import React, { Component } from 'react'
import fetch from 'isomorphic-fetch'

const fetchStates = ['loading', 'fetched', 'error']

export default class Fetch extends Component {
  constructor() {
    super()

    this.state = {
      fetchState: 'loading',
      data: null,
      error: null,
      loading: true,
      fetching: true,
      fetched: false
    }
  }

  async componentDidMount() {
    this.fetchData()
  }

  fetchData = async () => {
    const { url } = this.props

    try {
      const res = await fetch(url)
      const data = await res.json()

      this.setState({
        fetchState: 'fetched',
        editedData: data,
        data
      })
    } catch (error) {
      this.setState({
        error,
        fetchState: 'error'
      })
    }

    this.setState({
      loading: false,
      fetching: false,
      fetched: true
    })
  }

  handleRefetchClick = () => this.fetchData()

  handleDataChange = value => {
    if (this.state.fetching) {
      return
    }

    try {
      const data = JSON.parse(value)
      this.setState({ data, editedData: value })
    } catch (e) {
      this.setState({ editedData: value })
    }
  }

  handleFetchStateChange = e => {
    if (this.state.fetching) {
      return
    }

    const { value } = e.target

    this.setState({
      fetchState: value,
      loading: value === 'loading',
      error: value === 'error'
    })
  }

  render = () => this.props.children({
    fetchStates,
    onRefetchClick: this.handleRefetchClick,
    onDataChange: this.handleDataChange,
    onFetchStateChange: this.handleFetchStateChange,
    ...this.state
  })
}
