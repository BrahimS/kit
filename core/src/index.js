export {
  Library,
  Example,
  Detail
} from './Library'
export { default as LiveEditor } from './LiveEditor'
export { default as Frame } from './Frame'
export { default as Catch } from './Catch'
export { default as XRay } from './XRay'
export { default as PropsForm } from './PropsForm'
export { default as Responsive } from './Responsive'
export { default as Cartesian } from './Cartesian'
export { default as Matrix } from './Matrix'
export { default as Diff } from './Diff'
export { default as Debug } from './Debug'
export { default as TypeScale } from './TypeScale'
export { default as Style } from './Style'
export { default as Font } from './Font'
export { default as UI } from './ui'
