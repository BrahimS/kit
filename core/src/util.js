import { map, pipe, xprod, unnest, reduce, mergeAll } from 'ramda'

const arr = v => (Array.isArray(v) ? v : [v])
const log = msg => process.env.VERBOSE && console.log(msg)
const toSrcPath = (docsDir, path) =>
  path
    .replace(/\md$/, 'js')
    .replace(docsDir, 'src')
    .replace('components/', '')

const displayObj = obj =>
  Object.keys(obj)
    .map(k => `${k}=${obj[k]}`)
    .join(',')

// Adapted from: https://github.com/sindresorhus/titleize/blob/master/index.js
const titleize = str =>
  str.replace(/(?:^|\s|-)\S/g, m => m.toUpperCase()).replace(/(-|_)/g, ' ')

const isIndex = path => /index\.md/.test(path)
const introPage = nav =>
  nav[ROOT_LEVEL_FILE] &&
  nav[ROOT_LEVEL_FILE].find(n => n.name === 'introduction')

const extendDefaultProps = (Component, props) => {
  Component.defaultProps = Object.assign(
    {},
    Component.defaultProps || {},
    props
  )
}

const cartesianProduct = ({ theme = {}, ...props }) => {
  const xproduct = reduce(pipe(xprod, map(unnest)), [[]])

  const parsedProps = Object.keys(props).reduce((acc, k) => {
    return acc.concat([arr(props[k]).map(v => ({ [k]: v }))])
  }, [])

  return map(mergeAll, xproduct(parsedProps))
}

export {
  log,
  introPage,
  isIndex,
  titleize,
  toSrcPath,
  displayObj,
  extendDefaultProps,
  cartesianProduct
}
