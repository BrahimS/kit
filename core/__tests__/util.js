import { flatten } from '../src/util'

test('flatten flattens an object', () => {
  const result = flatten({
    red: 'red',
    gray: {
      light: ['#fafafa', '#f5f5f5'],
      dark: ['#555', '#111']
    },
    brand: {
      black: 'black',
      white: 'white',
      tomato: 'tomato'
    }
  })

  expect(result).toMatchSnapshot()
})
