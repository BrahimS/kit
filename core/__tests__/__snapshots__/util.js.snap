// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flatten flattens an object 1`] = `
Object {
  "brand.black": "black",
  "brand.tomato": "tomato",
  "brand.white": "white",
  "gray.dark.0": "#555",
  "gray.dark.1": "#111",
  "gray.light.0": "#fafafa",
  "gray.light.1": "#f5f5f5",
  "red": "red",
}
`;
